### PR TITLE
Add new package: Xyce

### PR DESCRIPTION
**Pull request by kuberry**
Adds a recipe for building Xyce and all dependencies.

### CODE REVIEW

File: var/spack/repos/builtin/packages/xyce/package.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    depends_on('trilinos@12.12.1~adios2~alloptpkgs+amesos+amesos2+anasazi+aztec+belos~boost~cgns~chaco+complex~cuda~cuda_rdc~debug~dtk+epetra+epetraext~exodus+explicit_template_instantiation~float+fortran~glm~gtest+hdf5~hwloc~hypre+ifpack~ifpack2~intrepid~intrepid2~ipo+isorropia+kokkos~matio~mesquite~metis~minitensor~ml~mpi~muelu~mumps~netcdf+nox~openmp~phalanx~piro~pnetcdf~python~rol~rythmos+sacado~shards~shared~shylu~stk~stratimikos~strumpack+suite-sparse~superlu~superlu-dist~teko~tempus+teuchos+tpetra+trilinoscouplings~wrapper~x11~xsdkflags~zlib+zoltan~zoltan2+stokhos+amesos2basker+epetraextbtf+epetraextexperimental+epetraextgraphreorderings gotype=\'none\'', when="~mpi")
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/xyce/package.py:55]: Why are we getting _this_ specific? Would this package not build if I use e.g. `trilinos+adios2`?

File: var/spack/repos/builtin/packages/xyce/package.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+    def cmake_args(self):
+        spec = self.spec
+
+        trilinos = spec['trilinos']
+
+        cxx_flags = [self.compiler.cxx_pic_flag]
+        try:
+            cxx_flags.append(self.compiler.cxx11_flag)
+        except:
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/xyce/package.py]: This is rejected by flake8 since it's too generic.

File: var/spack/repos/builtin/packages/xyce/package.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+class Xyce(CMakePackage):
+    """Xyce (zīs, rhymes with “spice”) is an open source, SPICE-compatible,
+    high-performance analog circuit simulator, capable of solving extremely
+    large circuit problems by supporting large-scale parallel computing
+    platforms. It also supports serial execution on all common desktop platforms,
+    and small-scale parallel runs on Unix-like systems. In addition to analog
+    electronic simulation, Xyce has also been used to investigate more general
+    network systems, such as neural networks and power grids.
+    """
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/xyce/package.py]: If you split the docstring in a first short paragraph followed by a longer description, e.g. like:
```suggestion
    """Xyce (zīs, rhymes with “spice”) is an open source, SPICE-compatible,
    high-performance analog circuit simulator, capable of solving extremely
    large circuit problems by supporting large-scale parallel computing
    platforms. 
    
    It also supports serial execution on all common desktop platforms,
    and small-scale parallel runs on Unix-like systems. In addition to analog
    electronic simulation, Xyce has also been used to investigate more general
    network systems, such as neural networks and power grids.
    """
```
Spack will use the short paragraph wherever a brief description is needed and the entire docstring instead in contexts where a longer description fits better.

**Review by alalazo — CHANGES_REQUESTED** [review]
Thanks for the contribution! I left a few comments below.

File: var/spack/repos/builtin/packages/xyce/package.py (hunk 1 above)
Comment (kuberry) [reply var/spack/repos/builtin/packages/xyce/package.py:55]: Unfortunately, it is necessary to be this specific at this time. A little background:
- Trilinos has a very large number of packages available, most of which would not be built by default in CMake
     - This means enabling any given package will require a variant
- For the users of the Trilinos recipe to-date (xSDK, Deal.ii, DTK), they have some packages that they like to enable by default, but these are not the packages that we would enable
    - This means unless we explicitly disable certain variants, then we will build many things we don't need

There are future plans (Fall), to move towards a Trilinos recipe that will disable all packages by default, and then users will explicitly enable the ones they want (like in CMake). However, even that will not result in a much less specific recipe (it will just remove all of the ~variants).

Comment (alalazo) [reply var/spack/repos/builtin/packages/xyce/package.py:55]: > This means unless we explicitly disable certain variants, then we will build many things we don't need

This is perfectly fine, at a package level, since the package is a place where requirements (not preferences) are stated. I understand that spending time to build unnecessary things is disappointing, but you can avoid that by customizing defaults in `packages.yaml`. Think of it the other way around: people that already installed e.g. trilinos with everything will need to install it again to match the required absence of some components here.

TL;DR can we remove already the `~variant`?

Comment (kuberry) [reply var/spack/repos/builtin/packages/xyce/package.py:55]: I can definitely appreciate what your are saying. And I think for most packages this would be the correct way to go. However, in this particularly case, Xyce requires the user to use version 12.12.1 of Trilinos, which is now several years old. There is a near 0% likelihood that the user will already have this version of Trilinos installed on their machine, unless it was specifically built to be used with Xyce. The result is that by removing the ~variants, we would require additional unneeded packages to be installed for any user building Xyce, with no reuse of the Trilinos package taking place since no user is likely to have that version of Trilinos with additional variants enabled.

Comment (kuberry) [reply var/spack/repos/builtin/packages/xyce/package.py:55]: Is there a way to disable the variant unless the user has **already** built that package with the variant? Basically, if we don't put ~variant, then it may be built and we don't need it. However, if it is already built, then we can use it (I don't think this exists in Spack, but I'm also new to Spack).

Also, Trilinos will update their recipe later this year to remove the needed for ~variants, but that is not yet the case.

Comment (alalazo) [reply var/spack/repos/builtin/packages/xyce/package.py:55]: > There is a near 0% likelihood that the user will already have this version of Trilinos installed on their machine, unless it was specifically built to be used with Xyce.

Thanks for pointing this out, it was not clear to me. Can you then add a code comment on top of the `depends_on` directives, explaining briefly what you stated in the two comments above? The main point is to document to an occasional reader that this is not the usual way other packages specify dependencies. Maybe also add a TODO pointing at a future removal of the `~variant`.

Comment (kuberry) [reply var/spack/repos/builtin/packages/xyce/package.py:55]: I've added that comment to the PR. Thanks for the suggestion to add the explanation.

**Review by alalazo — DISMISSED** [review]

**Comment by kuberry** [comment]
Had to remove the zis (with a long i). CentOS seemed to be choking on it and viewed it as a syntax error.

**Review by alalazo — APPROVED** [review]
